handler/effectiveschedule.go

```diff
@@ -0,0 +1,158 @@
+package handler
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	k8upv1alpha1 "github.com/vshn/k8up/api/v1alpha1"
+	"github.com/vshn/k8up/cfg"
+)
+
+// cleanupEffectiveSchedules removes elements in the EffectiveSchedule list that match the job type, but aren't randomized.
+// This is needed in case the schedule spec has changed from randomized to standard cron syntax.
+func (s *ScheduleHandler) cleanupEffectiveSchedules(jobType k8upv1alpha1.JobType, newSchedule k8upv1alpha1.ScheduleDefinition) {
+	es, found := s.effectiveSchedules[jobType]
+	if !found {
+		return
+	}
+	var schedules []k8upv1alpha1.ScheduleRef
+	for _, ref := range es.Spec.EffectiveSchedules {
+		if s.schedule.IsReferencedBy(ref) && !newSchedule.IsRandom() {
+			s.Log.V(1).Info("removing from effective schedule", "type", jobType, "schedule", k8upv1alpha1.GetNamespacedName(s.schedule))
+			continue
+		}
+		schedules = append(schedules, ref)
+	}
+	es.Spec.EffectiveSchedules = schedules
+	s.effectiveSchedules[jobType] = es
+}
+
+// synchronizeEffectiveSchedulesResources ensures that the effective schedules are created, updated or deleted depending on the Spec.
+// If no effective Schedule references the reconcile schedule resource anymore, it will be deleted.
+// On errors, the Ready condition will be set to false.
+func (s *ScheduleHandler) synchronizeEffectiveSchedulesResources() error {
+	for _, es := range s.effectiveSchedules {
+		if len(es.Spec.EffectiveSchedules) <= 0 {
```

Comment:
```suggestion
		if len(es.Spec.EffectiveSchedules) == 0 {
```

`len()` can not go below 0, can it?